### PR TITLE
feat: add prefix to all injected metrics

### PR DESCRIPTION
**Pull request by MoshPe**
This feat is about adding a prefix to all metrics that is being injected.
```js
PrometheusModule.register({
      prefix: 'prefix_',
      defaultMetrics: {
        config: {
          prefix: 'prefix_',
        },
        enabled: true,
      },
    }),
```
And the metrics will be:
```text/plain
prefix_nodejs_gc_duration_seconds_sum{kind="major"} 0.004125694999471307
prefix_nodejs_gc_duration_seconds_count{kind="major"} 1

# HELP prefix_valid_req_qty valid_req_qty
# TYPE prefix_valid_req_qty counter
prefix_valid_req_qty 0

# HELP prefix_valid_req_qty_metric valid_req_qty_metric
# TYPE prefixvalid_req_qty_metric counter
prefix_valid_req_qty_metric 0
```

### CODE REVIEW

File: .husky/pre-commit [hunk 1]
```diff
@@ -1,4 +1,6 @@
 #!/bin/sh
 . "$(dirname "$0")/_/husky.sh"
 
-npx --no-install lint-staged
+#npx --no-install lint-staged
+
+echo "lint-staged"
```
Comment (willsoto) [inline .husky/pre-commit:6]: Undo this please

File: package.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "@willsoto/nestjs-prometheus",
-  "version": "5.3.0",
-  "description": "NestJS module for Prometheus",
+  "version": "5.3.1",
```
Comment (willsoto) [inline package.json:3]: Undo this. Version is incremented automatically based on commit message

File: package.json [hunk 3]
```diff
@@ -1,7 +1,7 @@
...
+  "description": "NestJS module for Prometheus ",
```
Comment (willsoto) [inline package.json:4]: Undo

File: package.json [hunk 4]
```diff
@@ -12,12 +12,12 @@
     "typings"
   ],
   "scripts": {
-    "prebuild": "pnpm run clean",
+    "prebuild": "npm run clean",
```
Comment (willsoto) [inline package.json:15]: Undo all the `pnpm` -> `npm` changes

File: src/metrics/counter.ts [hunk 5]
```diff
@@ -10,8 +12,10 @@ export function makeCounterProvider(
 ): Provider {
   return {
     provide: getToken(options.name),
-    useFactory(): client.Metric<string> {
+    useFactory(config: PrometheusPrefix): client.Metric<string> {
```
Comment (willsoto) [inline src/metrics/counter.ts:15]: I would rather you just pass through the entire `PrometheusOptions` that already exists under the `PROMETHEUS_OPTIONS` symbol. The name you chose is a misnomer because you are only passing the prefix through.

File: src/metrics/counter.ts [hunk 6]
```diff
@@ -10,8 +12,10 @@ export function makeCounterProvider(
...
+      options.name = config.prefix.concat(options.name);
       return getOrCreateMetric("Counter", options);
```
Comment (willsoto) [inline src/metrics/counter.ts:17]: I would rather you update the signature of `getOrCreateMetric` to take an optional last arg of the options:

```ts
getOrCreateMetric(name: string, options, prometheusOptions?: PrometheusOptions)
```

And then add the prefixing logic there.

Comment (willsoto) [reply src/metrics/counter.ts:17]: Make it optional to preserve backwards compatibility

File: src/module.ts [hunk 7]
```diff
@@ -38,8 +38,13 @@ export class PrometheusModule {
         ),
       });
     }
+    providers.push({
+      provide: PROM_CONFIG,
+      useValue: {prefix: options?.prefix || ''},
+    })
 
     return {
+      global: true,
```
Comment (willsoto) [inline src/module.ts:47]: delete? Not sure why this was added

File: src/module.ts [hunk 8]
```diff
@@ -52,11 +57,20 @@ export class PrometheusModule {
     const controller = options.controller ?? PrometheusController;
 
     return {
+      global: true,
```
Comment (willsoto) [inline src/module.ts:60]: Delete.

File: test/metrics/gauge.spec.ts [hunk 9]
```diff
@@ -10,6 +11,12 @@ describe("Gauge", function () {
   beforeEach(async function () {
     testingModule = await Test.createTestingModule({
       providers: [
+        {
+          provide: PROM_CONFIG,
+          useValue: {
+            prefix: ""
+          }
+        },
```
Comment (willsoto) [inline test/metrics/gauge.spec.ts:19]: This shouldn't need to be added if it's being setup correctly in the module

File: src/module.ts [hunk 10]
```diff
@@ -176,6 +199,7 @@ export class PrometheusModule {
     options?: PrometheusOptions,
   ): PrometheusOptionsWithDefaults {
     return {
+      prefix: "",
```
Comment (willsoto) [inline src/module.ts:202]: Why initialize this if it's optional?

**Review by willsoto — CHANGES_REQUESTED** [review]

File: test/metrics/gauge.spec.ts (hunk 9 above)
Comment (MoshPe) [reply test/metrics/gauge.spec.ts:19]: Correct, just that this test does not provides the module, just the provider `makeGaugeProvider`, and it accepts an injection with provide of `PROMETHEUS_OPTIONS`.

Adding some tests regarding to that prefix being added to the name of the metric.

File: src/module.ts (hunk 7 above)
Comment (MoshPe) [reply src/module.ts:47]: The global is added to inject the prefix into all providers throughout the service.

File: .husky/pre-commit [hunk 11]
```diff
@@ -2,3 +2,4 @@
 . "$(dirname "$0")/_/husky.sh"
 
 npx --no-install lint-staged
+
```
Comment (willsoto) [inline .husky/pre-commit:5]: Revert this new line too please

File: package.json [hunk 12]
```diff
@@ -50,7 +50,7 @@
       "@semantic-release/commit-analyzer",
       "@semantic-release/release-notes-generator",
       "@semantic-release/github",
-      "@semantic-release/npm",
+      "@semantic-release/pnpm",
```
Comment (willsoto) [inline package.json:53]: Undo this change please

File: package.json [hunk 13]
```diff
@@ -34,7 +34,7 @@
       "prettier --write"
     ],
     "*.{ts,js}": [
-      "npm run lint",
+      "pnpm run lint",
```
Comment (willsoto) [inline package.json:37]: This is fine to keep. Should be `pnpm` anyway

File: package-lock.json
Comment (willsoto) [inline package-lock.json:1]: Delete this file please

File: src/interfaces.ts [hunk 14]
```diff
@@ -100,3 +104,7 @@ export interface PrometheusAsyncOptions
     ...args: unknown[]
   ): Promise<PrometheusOptions> | PrometheusOptions;
 }
+
+export interface PrometheusPrefix {
+  prefix: string;
+}
```
Comment (willsoto) [inline src/interfaces.ts:110]: Delete now?

File: src/metrics/utils.ts [hunk 15]
```diff
@@ -20,13 +21,18 @@ export type Options =
...
 ): client.Metric<string> {
   const existingMetric = client.register.getSingleMetric(options.name);
 
   if (existingMetric) {
     return existingMetric;
   }
 
+  options.name = prometheusOptions?.prefix
+    ? prometheusOptions?.prefix.concat(options.name)
+    : options.name;
```
Comment (willsoto) [inline src/metrics/utils.ts:34]: Let's not mutate the input please. Create a new `options` object using the prefix if they provided one.

File: src/module.ts (hunk 7 above)
Comment (willsoto) [reply src/module.ts:47]: The consumer can also import the `PrometheusModule` - which is something they need to do already anyway in order to inject metrics for other modules. So lets drop the `global` since it aligns with current behavior.

File: test/metrics/gauge.spec.ts [hunk 16]
```diff
@@ -10,6 +12,10 @@ describe("Gauge", function () {
   beforeEach(async function () {
     testingModule = await Test.createTestingModule({
       providers: [
+        {
+          provide: PROMETHEUS_OPTIONS,
+          useValue: sinon.stub(),
+        },
```
Comment (willsoto) [inline test/metrics/gauge.spec.ts:18]: Can you add a test actually testing that the prefixing works?

**Comment by willsoto** [comment]
Thanks. Closed via 2375e42c4b40203e9d09917ce6dd91390bb913b8
